In [1]:
# preamble
from datetime import datetime
start = datetime.now()

## Step 1 > create two new columns: ```price_label``` and ```busiest_day``` in imported data set ```orders_products_merged```

In [2]:
#import libraries
import pandas as pd
import numpy as np
import os

ICpath = r'C:\Users\joseh\Instacart Basket Analysis'

#import orders_products_merged.pkl
df_ords_prods_merged = pd.read_pickle(os.path.join(ICpath, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [3]:
df_ords_prods_merged.shape

(32404859, 15)

In [4]:
df_ords_prods_merged.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  first_order_from_customer  product_id  \
0                     NaN                       True         196   
1                    15.0                      False         196   
2                    21.0                      False         196   
3                    29.0                      False         196   
4                    28.0                      False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [5]:
# drop the column '_merge' with information on the merge results in a prior exercise
df_ords_prods_merged.drop(columns = ['_merge'], inplace=True)

#verify the change took place
df_ords_prods_merged.shape

(32404859, 14)

In [6]:
#using loc() method, assign price labels to 3 different segments of the data set
# High: price > 15 | Low: price <= 5 | Mid: price >5 and <=15
df_ords_prods_merged.loc[df_ords_prods_merged['prices']>15, 'price_label'] = 'High-range product'

In [7]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices']>5) & 
                         (df_ords_prods_merged['prices']<=15), 'price_label'] = 'Mid-range product'

In [8]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices']<=5, 'price_label'] = 'Low-range product'

In [9]:
#double check the results
df_ords_prods_merged['price_label'].value_counts()

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_label, dtype: int64

#### verified that the sum of the 3 segments above is 32,404,859:
21,860,860 + 10,126,321 + 417,678 = 32,404,859

In [10]:
#use a different method with a for-loop to create a new column (list) that will be named 'busiest_day'
result=[]
for value in df_ords_prods_merged['order_dow']:
    if value == 0:
        result.append('Busiest day')
    elif value ==4:
        result.append('Least busy')
    else:
        result.append('Regular busy')

In [11]:
#verify the number of rows of the new list is the same as those of the imported data frame
len(result)

32404859

In [12]:
#assign the newly created list to create a new column in the data frame
df_ords_prods_merged['busiest_day']=result

In [13]:
#verify the results
df_ords_prods_merged['busiest_day'].value_counts()

Regular busy    22416875
Busiest day      6204182
Least busy       3783802
Name: busiest_day, dtype: int64

In [14]:
# verify the total number of rows equal those of the original data frame
22416875 + 6204182 + 3783802

32404859

In [15]:
#check the size and the presence of the new columns in the data frame
df_ords_prods_merged

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0          2539329        1             1          2                  8   
1          2398795        1             2          3                  7   
2           473747        1             3          3                 12   
3          2254736        1             4          4                  7   
4           431534        1             5          4                 15   
...            ...      ...           ...        ...                ...   
32404854   1320836   202557            17          2                 15   
32404855     31526   202557            18          5                 11   
32404856    758936   203436             1          2                  7   
32404857   2745165   203436             2          3                  5   
32404858   3093936   205420             1          4                 14   

          days_since_prior_order  first_order_from_customer  product_id  \
0                            NaN                       True         196   
1                           15.0                      False         196   
2                           21.0                      False         196   
3                           29.0                      False         196   
4                           28.0                      False         196   
...                          ...                        ...         ...   
32404854                     1.0                      False       43553   
32404855                     3.0                      False       43553   
32404856                     NaN                       True       42338   
32404857                    15.0                      False       42338   
32404858                     NaN                       True       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices        price_label   busiest_day  
0               77              7     9.0  Mid-range product  Regular busy  
1               77              7     9.0  Mid-range product  Regular busy  
2               77              7     9.0  Mid-range product  Regular busy  
3               77              7     9.0  Mid-range product    Least busy  
4               77              7     9.0  Mid-range product    Least busy  
...            ...            ...     ...                ...           ...  
32404854        64              7     3.7  Low-range product  Regular busy  
32404855        64              7     3.7  Low-range product  Regular busy  
32404856        50             19     6.9  Mid-range product  Regular busy  
32404857        50             19     6.9  Mid-range product  Regular busy  
32404858       130             14    10.3  Mid-range product    Least busy  

[32404859 rows x 16 columns]

## Step 2 > Suppose your clients have changed their minds about the labels you created in your ```busiest_day``` column. Now, they want ```Busiest day``` to become ```Busiest days``` (plural). This label should correspond with the _two busiest days of the week_ as opposed to the single busiest day. At the same time, they’d also like to know the _two slowest days_. Create a new column for this using a suitable method.

In [16]:
df_ords_prods_merged['order_dow'].value_counts()

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_dow, dtype: int64

#### the two busiest days are 0, 1
#### the two slowest days are 3,4

In [17]:
df_ords_prods_merged.loc[  (df_ords_prods_merged['order_dow']==0)
                         | (df_ords_prods_merged['order_dow']==1), 'Busiest_Days'] = 'busiest days'

In [18]:
df_ords_prods_merged.loc[  (df_ords_prods_merged['order_dow']==3)
                         | (df_ords_prods_merged['order_dow']==4), 'Busiest_Days'] = 'slowest days'

In [19]:
df_ords_prods_merged.loc[(  df_ords_prods_merged['order_dow']==2)
                         | (df_ords_prods_merged['order_dow']==5)
                         | (df_ords_prods_merged['order_dow']==6), 'Busiest_Days'] = 'regular busy'

## Step 3 > Check the values of this new column for accuracy. Note any observations in markdown format.

In [20]:
df_ords_prods_merged['Busiest_Days'].value_counts()

regular busy    12916111
busiest days    11864412
slowest days     7624336
Name: Busiest_Days, dtype: int64

In [21]:
#verifying the number of rows remains consistent
12916111 + 11864412 + 7624336

32404859

## Step 4 > When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled 
* ```Most orders```
* ```Average orders```, and 
* ```Fewest orders``` 
### Create a new column containing these labels called ```busiest_period_of_day```


In [22]:
# verifying the frequency or orders on a given order
df_ords_prods_merged['order_hour_of_day'].value_counts()

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [23]:
# definition of ranges
# most orders -- from 8 to 19 hrs -- more than 1M orders in a given hour
# average orders -- from 0 to 1, from 6 to 7, and from 20 to 23 hrs --  between 100K and 1M in a given hour
# fewest orders -- from 2 to 5 hrs --less than 100K orders in given hour

In [24]:
# define a new list for the new field
orders_frequency=[]

for hour in df_ords_prods_merged['order_hour_of_day']:
    if (hour < 0) | (hour > 23): #out of range
        orders_frequency.append('Invalid entry')
    elif (hour >= 8) & (hour <= 19):
        orders_frequency.append('Most orders')
    elif (hour >= 2) & (hour <= 5):
        orders_frequency.append('Fewest orders')
    else:
        orders_frequency.append('Average orders')

In [25]:
#verifying the length to be the same as the number of records in data set
len(orders_frequency)

32404859

In [26]:
#create a new column/field in the data set
df_ords_prods_merged['busiest_period_of_day']=orders_frequency

In [27]:
df_ords_prods_merged

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0          2539329        1             1          2                  8   
1          2398795        1             2          3                  7   
2           473747        1             3          3                 12   
3          2254736        1             4          4                  7   
4           431534        1             5          4                 15   
...            ...      ...           ...        ...                ...   
32404854   1320836   202557            17          2                 15   
32404855     31526   202557            18          5                 11   
32404856    758936   203436             1          2                  7   
32404857   2745165   203436             2          3                  5   
32404858   3093936   205420             1          4                 14   

          days_since_prior_order  first_order_from_customer  product_id  \
0                            NaN                       True         196   
1                           15.0                      False         196   
2                           21.0                      False         196   
3                           29.0                      False         196   
4                           28.0                      False         196   
...                          ...                        ...         ...   
32404854                     1.0                      False       43553   
32404855                     3.0                      False       43553   
32404856                     NaN                       True       42338   
32404857                    15.0                      False       42338   
32404858                     NaN                       True       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices        price_label   busiest_day  \
0               77              7     9.0  Mid-range product  Regular busy   
1               77              7     9.0  Mid-range product  Regular busy   
2               77              7     9.0  Mid-range product  Regular busy   
3               77              7     9.0  Mid-range product    Least busy   
4               77              7     9.0  Mid-range product    Least busy   
...            ...            ...     ...                ...           ...   
32404854        64              7     3.7  Low-range product  Regular busy   
32404855        64              7     3.7  Low-range product  Regular busy   
32404856        50             19     6.9  Mid-range product  Regular busy   
32404857        50             19     6.9  Mid-range product  Regular busy   
32404858       130             14    10.3  Mid-range product    Least busy   

          Busiest_Days busiest_period_of_day  
0         regular busy           Most orders  
1         slowest days        Average orders  
2         slowest days           Most orders  
3         slowest days        Average orders  
4         slowest days           Most orders  
...                ...                   ...  
32404854  regular busy           Most orders  
32404855  regular 

## Step 5 > Print the frequency for this new column

In [28]:
df_ords_prods_merged['busiest_period_of_day'].value_counts()

Most orders       27818650
Average orders     4324350
Fewest orders       261859
Name: busiest_period_of_day, dtype: int64

In [29]:
#verifying total number of entries
27818650 + 4324350 + 261859

32404859

## Step 7 > Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder

In [30]:
df_ords_prods_merged.to_pickle(os.path.join(ICpath,'02 Data','Prepared Data','orders_products_derived_fields.pkl'))

In [31]:
# fyi only - to track performance of this notebook 
print(' time to run this notebook h:mm:ss: ', datetime.now() - start)

 time to run this notebook h:mm:ss:  0:00:41.719077


In [32]:
df_ords_prods_merged.shape

(32404859, 18)